### PR TITLE
Improve coverage checks

### PR DESCRIPTION
**Pull request by l0b0**
**Related Issue(s):** #331

**Description:** Add tests, exclude some lines from coverage count, include tests in coverage count (highlighting unreached test code), and do some minor code changes to make it more testable.

The first PR already contains a bunch of orthogonal improvements. Let me know if you want to split it up into simpler pieces.

**PR Checklist:**

- [x] Code is formatted (run `pre-commit run --all-files`)
- [x] Tests pass (run `scripts/test`)
- [ ] Documentation has been updated to reflect changes, if applicable
- [x] This PR maintains or improves overall codebase code coverage.
- [ ] Changes are added to the [CHANGELOG](https://github.com/stac-utils/pystac/blob/develop/CHANGELOG.md). See [the docs](https://pystac.readthedocs.io/en/latest/contributing.html#changelog) for information about adding to the changelog.

### CODE REVIEW

**Review by duckontheweb — COMMENTED** [review]
This looks good to me aside from using `TestCases.get_path` in some places.

I'm not sure what is going on with the Codecov.io reports. When I get the coverage report locally I get 93.17% coverage, but the report that's getting posted here shows 88.52% for this PR. I'll look into that separately and see if I can figure out the cause.

**Comment by duckontheweb** [comment]
> I'm not sure what is going on with the Codecov.io reports. When I get the coverage report locally I get 93.17% coverage, but the report that's getting posted here shows 88.52% for this PR. I'll look into that separately and see if I can figure out the cause.

@l0b0 In #452 I created a separate `coverage` CI job for uploading to Codecov.io so that we could get coverage reports even when coverage drops to help us identify gaps (this happened specifically in 71f07d93ce9cf219d09841d3dd3a55cdc690d066). I did this by directly using the `coverage run...` command in the new job, but I'm realizing now that I probably should have updated `scripts/test` to allow us to override `--fail-under` instead so that we wouldn't have to update that command in 2 places. 

I think the quick fix is to update the `coverage` CI job with the same changes you made to `scripts/test`. If you want to also update `scripts/test` so that we can use it in both cases that's great too 😄 , but I can do that separately.

File: pystac/extensions/sar.py [hunk 1]
```diff
@@ -318,7 +318,7 @@ def ext(cls, obj: T, add_if_missing: bool = False) -> "SarExtension[T]":
             return cast(SarExtension[T], AssetSarExtension(obj))
         else:
             raise pystac.ExtensionTypeError(
-                f"SAR extension does not apply to type {type(obj)}"
+                f"SAR extension does not apply to type '{type(obj).__name__}'"
```
Comment (duckontheweb) [inline pystac/extensions/sar.py:321]: It looks like this change could also be made for a number of other extensions as well. Is there a reason to only change this for the `sar` and `view` extensions, or should we update the `ext` method for all extensions in this PR?

Comment (l0b0) [reply pystac/extensions/sar.py:321]: I can do that, but I'd need to be sure that you're OK with changing all of them so it's not wasted effort. What do you think?

**Comment by l0b0** [comment]
Update: I'm splitting this into several smaller PRs for easier review/merging.

**Pull request closed without merging**